File: class-05/demos/linked-list.js
'use strict';

console.log('lets have some fun with a linked list');

function LinkedList(){
  this.head = null;
  this.tail = null;
}

LinkedList.prototype.insert = function(valuePotato){
  if(this.head === null) {
    const newNode = new Node(valuePotato);
    this.head = newNode;
    this.tail = newNode;
  } else {
    const newNode = new Node(valuePotato);
    const secondNode = this.head;
    this.head = newNode;
    newNode.next = secondNode;
  }

  // const newNode = new Node(valuePotato);
  // newNode.next = this.head;
  // this.head = newNode;
  // if(!this.tail) this.tail = newNode;
};

LinkedList.prototype.toStringIterative = function(){
  let str = '';
  let currentNode = this.head;
  while(currentNode !== null){
    str += currentNode.value + ',';
    currentNode = currentNode.next;
  }
  str += 'null';
  return str;
};

// Recursion: a function that calls itself
LinkedList.prototype.recursiveToString = function(){
  return this._recursiveToString(this.head);
};

LinkedList.prototype._recursiveToString = function (nodeWeAreOn) {

  // base case: when the problem is the smallest it will ever be
  // if the node is null, return "null"
  // work
  //
  // recursive part (when we call our own function);
  if(nodeWeAreOn === null) return 'null';
  return '' + nodeWeAreOn.value + ' -> ' + this._recursiveToString(nodeWeAreOn.next);
};

function Node(value) {
  this.value = value; // this should be of type int or String for today's lab
  this.next = null;
}

const linky = new LinkedList();

linky.insert('victor');
linky.insert('amelia');
console.log(linky);
console.log(linky.toStringIterative());
console.log(linky.recursiveToString());

File: class-06/code-review/dsa/challenges/src/main/java/challenges/linkedList/LinkedList.java
package challenges.linkedList;



public class LinkedList {
//    It knows about the head
    Node head; // == null

//    public LinkedList(){
//
//    }

    public void insert(int value){
        this.head = new Node(value, this.head);
    }

    public boolean includesWithALoop(int valueWeAreSearchingFor){
//         assuming LL is {9} -> {7} -> {5}
        Node current = this.head; // {value: 9}
        while(current != null) {
            if(current.value == valueWeAreSearchingFor) return true;
            current = current.next;
        }
        return false;
    }

    public boolean includesRecursive(int valueWeAreSearchingFor){
        return includesRecursive(this.head, valueWeAreSearchingFor);
    }

    private boolean includesRecursive(Node current, int valueWeAreSearchingFor){
//        assuming LL is {9} -> {7} -> {5} -> null
//        base case : we arrived at null : return false
//        small bit of work: node of not null : check for match: return true
//        recursion: call the function on current.next, value
        if(current == null) return false;
        if(current.value == valueWeAreSearchingFor) return true;
        return includesRecursive(current.next, valueWeAreSearchingFor);
    }
}

class Node {
    Node next;
    int value;

    public Node(int valuePotato){
        this.value = valuePotato;
    }

    public Node(int valuePotato, Node next){
        this.value = valuePotato;
        this.next = next;
    }
}

